app/ide-desktop/lib/dashboard/src/layouts/AssetProjectSessions.tsx

```diff
@@ -0,0 +1,62 @@
+/** @file A list of previous versions of an asset. */
+import * as React from 'react'
+
+import * as asyncEffectHooks from '#/hooks/asyncEffectHooks'
+
+import * as backendProvider from '#/providers/BackendProvider'
+
+import AssetProjectSession from '#/layouts/AssetProjectSession'
+
+import StatelessSpinner from '#/components/StatelessSpinner'
+import * as statelessSpinnerModule from '#/components/StatelessSpinner'
+
+import type * as backendModule from '#/services/Backend'
+
+import type AssetTreeNode from '#/utilities/AssetTreeNode'
+
+// =================
+// === Constants ===
+// =================
+
+/** The size (both width and height) of the loading spinner. */
+const SPINNER_SIZE = 32
+
+// ============================
+// === AssetProjectSessions ===
+// ============================
+
+/** Props for a {@link AssetProjectSessions}. */
+export interface AssetProjectSessionsProps {
+  readonly item: AssetTreeNode<backendModule.ProjectAsset>
+}
+
+/** A list of previous versions of an asset. */
+export default function AssetProjectSessions(props: AssetProjectSessionsProps) {
```

Comment:
Please, avoid default exports. They are harmfull

Reply:
addressing each point individually:
- poor discovery: not true, autocomplete picks up default exports with the correct name
  - developers can easily see the exported members of a module: a non-issue. generally react components only export a single component.
- refactoring: yes, this *is* an issue
- auto-complete: as above, works out of the box
- inconsistent codebases: technically true. however:
  - named imports can also trivially be renamed
  - again - auto-imports make this a non-issue in practice.
- re-exporting: only relevant when it is a library. also they missed out on:
```ts
import add from './add';
export { add };
```

tl;dr: all but one of the points simply do not hold up - many of the comments also mirror this sentiment.

even the React official docs prefer default exports: <https://react.dev/learn/referencing-values-with-refs#example-building-a-stopwatch>

so in this case i'd rather follow convention than switch conventions completely for little to no advantage

Reply:
> named imports can also trivially be renamed

It's not an issue since it's explicit.

> re-exporting: only relevant when it is a library

True, but we use reexporting a lot already

-- 

In the current codebase default export looks more like a bypass of the `import *` rule, don't you think?

this inconsistency leads to a problem when I don't know how to import a function/component upfront - do I need to import default? or should I need to use import * ?

> even the React official docs prefer default exports

They never said that they __prefer__ default exports, see: https://react.dev/learn/importing-and-exporting-components#exporting-and-importing-a-component

![CleanShot 2024-06-12 at 15 43 24@2x](https://github.com/enso-org/enso/assets/61194245/ca242e0b-3dde-49aa-b2c4-d09c8d1e57db)

--- 

All in all, in some cases, default exports make sense, but due to their `anonymous` and `implicit` nature, they're often forbidden. Too easy to make a mistake.

Reply:
> In the current codebase default export looks more like a bypass of the `import *` rule, don't you think?

no, as mentioned above it is because of React conventions. if not for that, i would rather namespace all components rather than using default exports...

> they never say that they *prefer* default exports.

true. however, *every single example in their docs* uses default exports. (the same applies to next.js docs.)

while it's true that export style tends to be divided mostly evenly between default and named exports, it's most certainly not a valid reason why named exports are better, or why default exports are worse.

> anonymous nature

not true, they take the names of the variable that is exported. it is *near impossible* to make a mistake because they, like everything else, are typically imported using auto-imports.

it's true that this may cause issues with things being out of sync when *renaming a file*. however this is an extremely rare occurrence and so I don't think it matters a lot

> we use reexporting a lot already

IMO re-exporting is mostly a non-concern when the component is only ever used in one place - i'm not convinced that they improve the organization of the code, not when the directory structure already offers the exact same level of organization

> this inconsistency leads to a problem when I don't know how to import a function/component upfront - do I need to import default? or should I need to use import * ?

public components were always default exports, so *this was a non-issue* until named exports were introduced - we have used default exports ever since the very beginning: <https://github.com/enso-org/enso/blob/6f29262f90027995efbe8d5eee6d831496e97ad7/app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/dashboard.tsx#L21>

Reply:
one final note is that it's worth noting that Vue also prefers default exports: <https://github.com/enso-org/enso/blob/0e17beba73884f06f352b92e20312712df7afdc5/app/gui2/src/components/GraphEditor/GraphNodes.vue#L2-L3>